scalableminds/webknossos
# Avoid allocating spire uint objects during apply agglomerate

When applying an agglomerate mapping, each value of the segmentation data needs to be converted to Long to then look up the agglomerate id in the hdf5 array. Since the jvm does not support unsigned types, this was previously done by wrapping each signed value in a spire unsigned integer object, and then calling toLong. 

This PR changes the code to skip using the spire classes and directly calling the to-long method (which performs bitwise and to do the unsigned logic) on each value.

It also moves the filter-zero logic out of the data conversion since this had a performance cost even if the filterZero boolean was false.

It also adds a pre-allocated LongBuffer instead of using `array.map(_.toLong)`, which is slightly faster, I guess the old version did not know the final length of the output array before.

My measurements show that the conversion to Long is about six times faster with this PR, almost halving the time to apply an agglomerate mapping on my test data.

cc @youri-k 

### URL of deployed dev instance (used for testing):
- https://applymappingfunctionaltyping.webknossos.xyz

### Steps to test:
- View dataset with agglomerate mapping; activate it, should show mapped data
- Compute ad-hoc mesh, should show meshes for mapped data
- Histogram and find data should still work for different dtypes (e.g. uint8, uint16, float, uint24-rgb)

------
